# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Create Path and Import New 'Customers' data

In [2]:
# Create path
path = "/Users/charlottelin/Documents/06-2025 Instacart Basket Analysis"

In [3]:
# Import the customers data
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
# Check dataset 
df_customers.shape

(206209, 10)

# Step 4 - Wrangle 'customers' Data

In [5]:
# Investigate column names
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

#### Observation: based on the output, I would take the following wrangling actions:
- standardize column names to all lowercase for consistency and readability
- make sure to remove all whitespace
- change the date_joined date format to a 'datetime' format that's more friendly for filtering

In [6]:
# Improve column consistency and readability
df_customers.rename(columns={
    'First Name': 'first_name',
    'Surname': 'surname',
    'Gender': 'gender',
    'STATE': 'state',
    'Age': 'age',
    'date_joined': 'date_joined',
    'n_dependants': 'num_dependents',
    'fam_status': 'marital_status',
    'income': 'income'
}, inplace=True)

In [7]:
# Change date format in date_joined
df_customers['date_joined'] = pd.to_datetime(df_customers['date_joined'])

In [8]:
# Check output
df_customers.head()

user_id first_name    Surnam  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48  2017-01-01   
1    33890   Patricia      Hart  Female  New Mexico   36  2017-01-01   
2    65803    Kenneth    Farley    Male       Idaho   35  2017-01-01   
3   125935   Michelle     Hicks  Female        Iowa   40  2017-01-01   
4   130797        Ann   Gilmore  Female    Maryland   26  2017-01-01   

   num_dependents marital_status  income  
0               3        married  165665  
1               0         single   59285  
2               2        married   99568  
3               0         single   42049  
4               1        married   40374

In [9]:
# Investigate Data Types
df_customers.dtypes

user_id                    int64
first_name                object
Surnam                    object
gender                    object
state                     object
age                        int64
date_joined       datetime64[ns]
num_dependents             int64
marital_status            object
income                     int64
dtype: object

#### Observation: 'user_id' is currently an integer; however, user ID is an identifier, not a quantity. Therefore, it is best to change it to a string to help avoid misinterpretation.

In [10]:
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [11]:
# Check output
df_customers['user_id'].dtype

dtype('O')

#### Wrangling conclusion:
- I reviewed all columns in the dataset and found each one relevant to the analysis objectives. Therefore, I will not be dropping any columns at this stage.
- Transposing the data is also unnecessary, as the current structure is well-suited for analysis.
- All columns are now named with consistency and have appropriate data types.

# Step 5 - Data Quality & Consistency Checks

#### <b>Summary Statistics</b>

In [12]:
# Investigate summary statistics
df_customers.describe()

age                    date_joined  num_dependents  \
count  206209.000000                         206209   206209.000000   
mean       49.501646  2018-08-17 03:06:30.029532928        1.499823   
min        18.000000            2017-01-01 00:00:00        0.000000   
25%        33.000000            2017-10-23 00:00:00        0.000000   
50%        49.000000            2018-08-16 00:00:00        1.000000   
75%        66.000000            2019-06-10 00:00:00        3.000000   
max        81.000000            2020-04-01 00:00:00        3.000000   
std        18.480962                            NaN        1.118433   

              income  
count  206209.000000  
mean    94632.852548  
min     25903.000000  
25%     59874.000000  
50%     93547.000000  
75%    124244.000000  
max    593901.000000  
std     42473.786988

#### Observation: the summary stats appear reasonable. I don't spot any outliers or unexpected ranges.

#### <b>Mixed-Type Data Check</b>

In [13]:
# Check for any columns with mixed-type data
for col in df_customers.columns.tolist():
    weird = (df_customers[[col]].applymap(type) != df_customers
             [[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_customers[weird]) > 0:
        print(col)

/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame

first_name


/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_2399/409413604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_customers[[col]].applymap(type) != df_customers


#### Observation: I will stop using applymap() on a dataframe for a smoother workflow with pandas. As 'first_name' was flagged, I will proceed to fix its data types.

In [14]:
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [15]:
# Check output
df_customers['first_name'].dtype

dtype('O')

In [16]:
# Conduct a second check with updated code
for col in df_customers.columns.tolist():
    weird = (df_customers[[col]].map(type) != df_customers
             [[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_customers[weird]) > 0:
        print(col)

#### Observation: This check returned nothing so we're good to proceed.

#### <b>Check for Missing Values</b>

In [17]:
df_customers.isnull().sum()

user_id           0
first_name        0
Surnam            0
gender            0
state             0
age               0
date_joined       0
num_dependents    0
marital_status    0
income            0
dtype: int64

#### Observation: There are no missing values in this dataframe.

#### <b>Check for Duplicates</b>

In [18]:
df_dups = df_customers[df_customers.duplicated()]

In [19]:
# Check output
df_dups.head()

Empty DataFrame
Columns: [user_id, first_name, Surnam, gender, state, age, date_joined, num_dependents, marital_status, income]
Index: []

#### Observation: There are no duplicate values in this dataframe.

# Step 6 - Combine Customer Data with Existing Prepared Data

In [20]:
# Import the latest prepared dataframe
df_prepared_dataframe = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4_8.pkl'))

In [21]:
# Check imported dataframe
df_prepared_dataframe.shape

(32404859, 25)

In [22]:
# Create a subset of the prepared dataframe to manage memory
subset_size = 10_000_000
df_subset = df_prepared_dataframe.sample(n=subset_size, random_state=42)

In [23]:
# Check output
df_subset.shape

(10000000, 25)

#### Note: this subset was created with random sampling, at approximately 1/3 of the original data size.

In [24]:
# Check 'user_id' data type in prepared dataframe
df_subset['user_id'].dtype

dtype('int64')

In [25]:
# Convert it to string to match df_customers
df_subset['user_id'] = df_subset['user_id'].astype('str')

In [26]:
# Check output
df_subset['user_id'].dtype

dtype('O')

In [27]:
# Check for unique values in prepared dataframe
df_subset['user_id'].nunique()

205562

In [28]:
# Do the same for df_customers
df_customers['user_id'].nunique()

206209

#### Observation: Although the `user_id` count in `df_subset` (205,562) is slightly lower than in `df_customers` (206,209) due to subsetting, the overlap is strong enough to proceed with the merge without significantly impacting data completeness or analysis quality.

#### <b>Merge the two dataframes</b>

In [29]:
#Perform a left merge to retain the full sample
df_merging = df_subset.merge(df_customers, on='user_id', how='left', indicator='merging_status')

In [30]:
# Check output
df_merging.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3411074  125899             1                   2                 20   
1    582767  131225             6                   4                  8   
2   1648217  129928            31                   1                 20   
3    492489   93812            17                   0                 12   
4    277503   67925             2                   0                 13   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN       27283                  3          0   both   
1                    28.0        3896                  3          1   both   
2                    10.0       37029                 47          1   both   
3                     8.0       18234                 22          1   both   
4                     2.0       41950                  3          0   both   

   ... first_name      Surnam  gender     state age date_joined  \
0  ...     Brenda  Pennington  Female  Oklahoma  22  2019-03-22   
1  ...     Victor       Burke    Male    Alaska  19  2019-05-05   
2  ...       Adam   Fernandez    Male  Oklahoma  61  2018-03-30   
3  ...      Nancy       Gomez  Female    Nevada  50  2017-08-08   
4  ...     Rachel      Torres  Female     Texas  25  2019-02-19   

  num_dependents                    marital_status  income  merging_status  
0              3                           married   32609            both  
1              3  living with parents and siblings   50072            both  
2              0                  divorced/widowed   36163            both  
3              0                            single  163081            both  
4              2                           married   41667            both  

[5 rows x 35 columns]

In [31]:
df_merging.shape

(10000000, 35)

In [32]:
df_merging['merging_status'].value_counts()

merging_status
both          10000000
left_only            0
right_only           0
Name: count, dtype: int64

#### Observation: the value.count indicates that the full merge was successful.

# Step 8 - Export Final Pickle

In [33]:
df_merging.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'combined_customers_4_9_Final.pkl'))